### PR TITLE
GPII-1785:  Update `gpii-pouchdb-lucene` to latest `gpii-express` and `gpii-pouchdb`...

### PR DESCRIPTION
**Pull request by the-t-in-rtf**
See [GPII-1785](https://issues.gpii.net/browse/GPII-1785) for details.

_NOTE:_ This PR is not ready for review yet, it should be reviewed only after [the associated PR for `gpii-pouchdb`](https://github.com/GPII/gpii-pouchdb/pull/10) is merged.

### CODE REVIEW

**Comment by the-t-in-rtf** [comment]
This is ready for review, but should not be merged/published until its dependencies are merged/published:

* [gpii-express](https://github.com/GPII/gpii-express/pull/25)
* [gpii-pouchdb](https://github.com/GPII/gpii-pouchdb/pull/12)

**Comment by the-t-in-rtf** [comment]
@amb26, now that gpii-express and gpii-pouchdb have been merged, this should be ready for review.

**Comment by amb26** [comment]
A mysterious effect of running the build in this branch on Windows is to create some strange kind of softlink from the "build" directory to the root of the overall drive - can we trace what is causing this behaviour?

E:\Source\gits\gpii-pouchdb-lucene>dir build
 Volume in drive E is FAST NTFS DATA
 Volume Serial Number is 3411-D59E

 Directory of E:\Source\gits\gpii-pouchdb-lucene\build

07/12/2016  18:24    <DIR>          .
07/12/2016  18:24    <DIR>          ..
07/12/2016  18:25    <DIR>          couchdb-lucene
07/12/2016  18:24    <DIR>          E?
07/12/2016  18:22               108 README.md
               1 File(s)            108 bytes
               4 Dir(s)  45,122,646,016 bytes free

File: bin/build_couchdb_lucene.js [hunk 1]
```diff
@@ -0,0 +1,149 @@
+/*
```
Comment (amb26) [inline bin/build_couchdb_lucene.js]: We could do with some comments explaining the purpose and strategy of this file - together with a few impl comments

File: src/js/gpii-pouchdb-lucene.js [hunk 2]
```diff
@@ -52,12 +52,13 @@ var os            = require("os");
 var child_process = require("child_process");
 var fs            = require("fs");
 
-// There are some constants that we use in `npm` scripts as well as here, they are found in `package.json`.
-var npmSettings = require("../../package.json");
+// require("../../");
```
Comment (amb26) [inline src/js/gpii-pouchdb-lucene.js]: This can be removed? I guess there are no other impl files

File: tests/js/lucene-tests.js [hunk 3]
```diff
@@ -1,5 +1,9 @@
 // Sanity checks for search integration.  As these are read only, we use a single instance for all tests.
 "use strict";
+
+var events = require("events");
+events.EventEmitter.defaultMaxListeners = 100;
```
Comment (amb26) [inline tests/js/lucene-tests.js]: Need a comment tracing what problem this works around - and if possible with an issue link for whatever upstream project the problem is with

File: src/js/gpii-pouchdb-lucene.js (hunk 2 above)
Comment (the-t-in-rtf) [inline src/js/gpii-pouchdb-lucene.js]: Done.

File: bin/build_couchdb_lucene.js (hunk 1 above)
Comment (the-t-in-rtf) [inline bin/build_couchdb_lucene.js]: I made a first pass at this, please review.

File: tests/js/lucene-tests.js (hunk 3 above)
Comment (the-t-in-rtf) [inline tests/js/lucene-tests.js]: The tests in this package no longer trigger the warning.  I have removed these lines for now.  If the problem crops up when running larger sets of tests in other packages (such as the UL API), we can revisit this there.

**Comment by the-t-in-rtf** [comment]
> A mysterious effect of running the build in this branch on Windows is to create some strange kind of softlink from the "build" directory to the root of the overall drive - can we trace what is causing this behaviour?

@amb26, I cannot reproduce the "build" directory oddities in a vagrant Windows VM.  I am running the build from c:\users\vagrant\gpii-pouchdb-lucene, also three levels deep, so it seems like something different than the usual problems with parent directory references.

**Comment by the-t-in-rtf** [comment]
I was reviewing the build script to confirm, it uses fluid.module.resolvePath("%gpii-pouchdb-lucene/build").  While I was there, I noticed and removed a redundant call to fluid.module.resolvePath, which shouldn't be relevant.  Both the build and tests still run fine in a Windows Vagrant VM with the changes.

**Comment by amb26** [comment]
This looks essentially ready to merge to me - I just have a couple of small queries - 
i) When I run a yarn build, I end up with some diffs in the yarn.lock file - they look pretty trivial apart from a change in the version of readable-stream - is this something that indicates that it is stale wrt package.json or is this something I can generally expect from yarn?
ii) Given that the Vagrant build is broken I imagine we should add a disclaimer to that effect
iii) perhaps build/couchdb-lucene should be added to .gitignore?

**Comment by the-t-in-rtf** [comment]
1. `yarn upgrade` uses relative versions by default.  I "pinned" the versions in the package.json, but hadn't rerun `yarn` or checked in the updated `yarn.lock` with the "pinned" versions.  It should be fixed now.
2. The Vagrant build works now, no need for a note.
3. Done.

**Comment by the-t-in-rtf** [comment]
@amb26, I have addressed all remaining feedback, this is ready for your review again.

**Comment by the-t-in-rtf** [comment]
Sorry, @amb26, spoke too soon.  The QI build appears to have passed, but actually did not.  I am verifying locally and then working with the QI team to get the build working.  I will advise once this is fixed.  FYI, [GPII-2133](https://issues.gpii.net/browse/GPII-2133) is a critical issue where we are discussing build failures that are incorrectly reported as passing.

**Comment by the-t-in-rtf** [comment]
OK, @amb26, the Vagrant tests succeed locally, I am working on the QI issues with the team.  I'm happy enough with the state of things here to proceed with the merge when you're ready.

**Comment by the-t-in-rtf** [comment]
@amb26, I just updated this one for further review.  Although there are still problems with getting the job to run in QI, these should not block the merge, as long as the `Vagrantfile` and `.qi.yml` can be used to run `vagrant up` and test locally (which they can).

**Comment by the-t-in-rtf** [comment]
In updating a package that uses gpii-pouchdb-lucene for its tests, I noticed failing tests, which led me to discover that the tests are not passing at all, at least on OS X.  I will need to investigate further.

@amb26, please hold off on this one until further notice.

**Comment by the-t-in-rtf** [comment]
I looked through recent changes and confirmed that [this was the last change before the tests started failing](https://github.com/the-t-in-rtf/gpii-pouchdb-lucene/commit/b3a067155cdf7a7cff10d4992e60fc55f50b8cb9).  I am looking at those changes now.

**Comment by the-t-in-rtf** [comment]
Looks like it was specifically the upgrade of gpii-pouchdb beyond version 1.0.1.  I will review those changes now.

**Comment by the-t-in-rtf** [comment]
This is the commit that broke things:

https://github.com/the-t-in-rtf/gpii-pouchdb/commit/e73a3e024893eb5e959c550a42c25fa917371a32

**Comment by the-t-in-rtf** [comment]
I found the culprit, and have written up the problem as [GPII-2422](https://issues.gpii.net/browse/GPII-2422).

**Comment by the-t-in-rtf** [comment]
@amb26, with the fix for GPII-2422 in place, all tests are again passing natively and in Vagrant.  This PR is ready for review again.

**Comment by the-t-in-rtf** [comment]
@amb26, I would like to close this out so that I can prepare a pull to update this library for compatibility with CouchDB 2.0.  As far as I can see, we're only a little bit of verification away from merging, we were quite close 8 months or so ago.

**Pull request merged**